crates/typos-cli/src/config.rs

```diff
@@ -19,26 +20,55 @@ pub struct Config {
     pub overrides: EngineConfig,
 }
 
+#[derive(Debug, Clone, Default, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+#[serde(default)]
+#[serde(rename_all = "kebab-case")]
+pub struct PyprojectTomlConfig {
+    pub tool: PyprojectTomlTool,
+}
+
+#[derive(Debug, Clone, Default, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+#[serde(default)]
+#[serde(rename_all = "kebab-case")]
+pub struct PyprojectTomlTool {
+    pub typos: Option<Config>,
+}
+
 impl Config {
     pub fn from_dir(cwd: &std::path::Path) -> Result<Option<Self>, anyhow::Error> {
-        let config = if let Some(path) = find_project_file(cwd, SUPPORTED_FILE_NAMES) {
-            log::debug!("Loading {}", path.display());
-            Some(Self::from_file(&path)?)
-        } else {
-            None
-        };
-        Ok(config)
+        for file in find_project_files(cwd, SUPPORTED_FILE_NAMES) {
+            log::debug!("Loading {}", file.display());
+            if let Some(config) = Self::from_file(&file)? {
+                return Ok(Some(config))
+            }
+        }
+
+        Ok(None)
     }
 
-    pub fn from_file(path: &std::path::Path) -> Result<Self, anyhow::Error> {
+    pub fn from_file(path: &std::path::Path) -> Result<Option<Self>, anyhow::Error> {
         let s = std::fs::read_to_string(path).map_err(|err| {
             let kind = err.kind();
             std::io::Error::new(
                 kind,
                 format!("could not read config at `{}`", path.display()),
             )
         })?;
-        Self::from_toml(&s)
+
+        if path.file_name().unwrap() == "pyproject.toml" {
+            return match toml::from_str::<PyprojectTomlConfig>(&s) {
+                Err(_) => {
+                    log::debug!("No `tool.typos` section found in `pyproject.toml`, skipping");
+                    Ok(None)
+                },
+                Ok(config) => Ok(config.tool.typos)
+            }
+        }
+
+        match Self::from_toml(&s) {
+            Ok(x) => Ok(Some(x)),
+            Err(x) => Err(x)
+        }
```

Comment:
nit: imo these are two mutually exclusive business-logic paths (not just bookkeeping) and that should be reflected in the control flow with an `else`